### PR TITLE
HW Sockets

### PR DESCRIPTION
**Pull request by UsernameAlexei**
HW Sockets, maximum task

### CODE REVIEW

File: TCP_client.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import socket
+import time
+from Data_collection_class import Collection
+from datetime import datetime
+
+sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Comment (arybin93) [inline TCP_client.py]: Context managers as well

**Review by arybin93 — COMMENTED** [review]
Thank you for your work. Looks good. I've left some comments, please resolve them and update PR.

File: Data_collection_class.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        threading.Thread.__init__(self, target=self.start_collect)
+        self.counter = 0
+        self.running = False
+
+    def start_collect(self):
+        self.running = True
+        not_pressed = True
+
+        while self.running:
+            keyboard.read_key()
```
[the remark this answers is not in the document]
Comment (UsernameAlexei) [reply Data_collection_class.py]: Maybe we can start ```stop_collect``` by pressing a specific key? Having previously added the condition to ```stop_collect```
```python    

def start_collect(self):
      self.running = True
      not_pressed = True
        
      keyboard.on_press(self.stop_collect)
      while self.running:
          keyboard.read_key()
          if not_pressed:
              self.counter += 1
              print(self.counter)
              not_pressed = False
          else:
              not_pressed = True

def stop_collect(self, event):
        if event.name =='q':
            self.running = False
            pyautogui.press('esc')
            self.counter -= 1
            threading.Thread.join(self)


```

File: keystrokes.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+['01:28:731867', '9 press the key in 1 min']
```
[the remark this answers is not in the document]
Comment (UsernameAlexei) [reply keystrokes.txt:1]: Thanks for the additional material. Uploaded the updated output keystrokes.csv

File: TCP_client.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    keystrokes = Collection()
+    keystrokes.cleanup()
+    keystrokes.start()
+    time.sleep(60)
+    keystrokes.stop_collect()
+    # time tracking
+    time_now = datetime.now()
+    time_now = time_now.strftime("%H:%M:%f")
+
+    result = [time_now, f'{keystrokes.get_current_state()} press the key in 1 min']
```
[the remark this answers is not in the document]
Comment (UsernameAlexei) [reply TCP_client.py]: Ready, updated.

File: TCP_server.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import socket
+
+sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
[the remark this answers is not in the document]
Comment (UsernameAlexei) [reply TCP_server.py]: Thank you.

File: Data_collection_class.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    def cleanup(self):
+        self.counter = 0
+
+    def stop_collect(self):
+        self.running = False
+        pyautogui.press('esc')
+        self.counter -= 1
+        threading.Thread.join(self)
+
+# a = Collection()
```
[the remark this answers is not in the document]
Comment (UsernameAlexei) [reply Data_collection_class.py]: Indeed, we must not forget about the tests. 
Uploaded test_collection.py

File: TCP_client.py (hunk 1 above)
Comment (UsernameAlexei) [reply TCP_client.py]: Ready, updated.

File: TCP_server.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    print('Connected: ', addr)
+
+    while True:
+        data = client.recv(1024)
+        if not data:
+            break
+        udata = data.decode()
+        client.send(udata.encode('utf-8'))
+        # save to file
+        with open('keystrokes.txt', 'a') as f:
```
[the remark this answers is not in the document]
Comment (UsernameAlexei) [reply TCP_server.py]: Use a mutex?

File: TCP_server.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (UsernameAlexei) [reply TCP_server.py]: Updated

File: TCP_server.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (arybin93) [reply TCP_server.py]: Yes, right

File: Data_collection_class.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arybin93) [reply Data_collection_class.py]: I mean something like that:
```
def handler(event):
    self.counter += 1

keyboard.on_press(handler) 
while self.running:
    # just waiting small time for saving CPU time
    time.sleep(0.01) 
```
And add `keyboard.unhook_all()` to `stop_collect()` for cancel handling

File: TCP_server.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import socket
+import csv
+import ast
+from contextlib import closing
+
+with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
```
Comment (arybin93) [inline TCP_server.py:6]: `with socket.socket() as s` should work as well

**Review by arybin93 — APPROVED** [review]
Looks better, thank you for your work. Homework is accepted.

File: Data_collection_class.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (UsernameAlexei) [reply Data_collection_class.py]: Updated. Thank you, this is a really very good solution.

**Comment by UsernameAlexei** [comment]
Thank you, the module is interesting, with clear examples in the lecture. Studying additional material.

**Pull request merged**